### PR TITLE
C2S docs

### PR DESCRIPTION
**Pull request by chrzaszcz**
Update documentation after introducing `mongose_c2s`.
The migration guide will be updated in a separate PR.

Main changes:
- New routing docs and diagrams, since the old ones were outdated. It was difficult to get any good-looking output from `plantuml`, so it was replaced with Google docs. The docs were looking bad in night mode, so a separate dark-mode version is used.
- Hook description is updated and unified. There were multiple updates since this doc was out of date.
- Missing configuration options are described.
- Hook API modules moved to `src/hooks` to make them easier to find and automatically grep. Another reason was naming - we had confusing modules like `mongoose_metric_hooks`, which contains handlers, not hooks. The rename is the simplest change, and I wanted to avoid code changes in this PR.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/165665) / small_tests / 6af368cdc43f9e84107fb0b37a1811b99a032800
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3962/165665/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3962/165665/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-02-13_18.14.59/lib.mongooseim.logs/run.2023-02-13_18.15.00/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/165773) / small_tests / c7a08ed6743c2f3913c77fb3ca897dc6fc62d275
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3962/165773/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3962/165773/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-02-14_17.33.05/lib.mongooseim.logs/run.2023-02-14_17.33.06/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/165899) / small_tests / b8425673ea8316112f49ef5566a2a88d9018d606
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3962/165899/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3962/165899/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-02-15_17.51.22/lib.mongooseim.logs/run.2023-02-15_17.51.23/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/165930) / small_tests / cb8d6752903f8637e7633aab7fdf19881f5f0159
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3962/165930/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3962/165930/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-02-15_18.23.18/lib.mongooseim.logs/run.2023-02-15_18.23.19/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/165953) / small_tests / 245c2792fad4247ac362f74361a0e5617b0331fc
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3962/165953/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3962/165953/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-02-16_08.49.28/lib.mongooseim.logs/run.2023-02-16_08.49.29/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_25](https://circleci.com/gh/esl/MongooseIM/166012) / small_tests / aa5188790461bc3805cbd11a79f7ad676aad092a
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3962/166012/small_tests.25.2) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3962/166012/small_tests.25.2/small/ct_run.mongooseim@localhost.2023-02-16_11.27.44/lib.mongooseim.logs/run.2023-02-16_11.27.45/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/166072) / small_tests / 7bbc91f51c5fd1d718c204e2aa4e5983169da627
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3962/166072/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3962/166072/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-02-16_18.04.32/lib.mongooseim.logs/run.2023-02-16_18.04.33/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/166094) / small_tests / 8c8c748c272e0c2f363f64342e344e0b26823e70
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3962/166094/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3962/166094/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-02-17_08.26.33/lib.mongooseim.logs/run.2023-02-17_08.26.34/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/166130) / small_tests / 7939b1310eb41ed905dbdc467e62a81927fb0a2f
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3962/166130/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3962/166130/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-02-17_08.33.26/lib.mongooseim.logs/run.2023-02-17_08.33.26/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/166132) / small_tests / 045d2ecdbda80fcba061684ccaf3de6b3b5b4686
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3962/166132/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3962/166132/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-02-17_08.33.57/lib.mongooseim.logs/run.2023-02-17_08.33.58/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/166176) / small_tests / 6d2c7740e559ada1993a2b22011cead1891a5807
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3962/166176/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3962/166176/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-02-17_09.20.13/lib.mongooseim.logs/run.2023-02-17_09.20.13/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/166181) / small_tests / a9dc040deecc5dc19d254ae06d51e7089aab0a29
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3962/166181/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3962/166181/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-02-17_09.22.14/lib.mongooseim.logs/run.2023-02-17_09.22.14/suite.log.html)

File: doc/developers-guide/Stanza-routing.md [hunk 1]
```diff
@@ -1,76 +1,74 @@
...
 
-## 8. `ejabberd_c2s`
+* `user_receive_message` for messages,
+* `user_receive_presence` for presences,
+* `user_receive_iq` for IQ (info/query),
+* `user_receive_xmlel` for other XML elements.
 
-`ejabberd_c2s` verifies the stanza against relevant privacy lists and sends it to the socket.
-`user_receive_packet` hook is run to notify the rest of the system about the stanza delivery to User B.
+Each hook can be handled by multiple modules subscribed to it. These hooks' handlers can stop the routing, e.g. when the stanza is blocked by `mod_privacy`. Finally, the `xmpp_presend_element` hook is called, which can stop the routing as well - otherwise, the stanza is converted to binary data and sent to the recipient's TCP socket.
```
Comment (NelsonVides) [inline doc/developers-guide/Stanza-routing.md:74]: I'd maybe add here that `xmpp_presend_element` is where stream management or csi hook in to handle buffering, just as an example 🤔

File: doc/developers-guide/Stanza-routing.md [hunk 2]
```diff
@@ -1,76 +1,74 @@
...
 
-## 3. Privacy lists and `ejabberd_router:route/3`
+* `user_send_message` for messages,
+* `user_send_presence` for presences,
+* `user_send_iq` for IQ (info/query) stanzas,
+* `user_send_xmlel` for other XML elements.
 
-The stanza is checked against any privacy lists in use and, in the case of being allowed, it will be routed by `ejabberd_router:route/3`. 
-This also takes into account "blocking commands", which are part of the privacy system.
+Each hook can be handled by multiple modules subscribed to it. Those modules do various complementary tasks, like storing the message in an archive, sending carbon copies, checking the stanza against privacy lists etc. See [hooks description](../hooks_description/#hooks-called-for-session_established) for more information.
```
Comment (NelsonVides) [inline doc/developers-guide/Stanza-routing.md:22]: I'd also add here that these hooks can completely halt the pipeline, if they return `{stop, NewAcc}`, we will not only stop the current hook as generally `gen_hook` does, but we will also not run the following one, meaning, a stop in `user_send_packet` will also avoid running the subsequent `user_send_*`.

**Review by NelsonVides — APPROVED** [review]
Two minor comments, dunno if those are actually useful examples but just throwing in some thought. Otherwise, a wonderful PR 🎉

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/166226) / small_tests / 87168574fbe0bf4e05482014d82a7008ce197d41
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3962/166226/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3962/166226/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2023-02-17_11.36.06/lib.mongooseim.logs/run.2023-02-17_11.36.06/suite.log.html)

**Review by NelsonVides — APPROVED** [review]
🎉

**Pull request merged**